### PR TITLE
orchagent: Adding QosOrch and BufferOrch classes

### PR DESCRIPTION
**Pull request by hrachya-mellanox**
* Class BufferOrch to handle buffer configuration
*   refactored Orch based class to contain common functions from QosOrch
and BufferOrch
* Class QosOrch
*   serialization for qos -
*	PFC-PG, PFC-TC, TC-PG, DSCP-TC, TC-queue maps
*	ECN support
*	WRED support
*		NOTE: Only max for green/yellow can specified. The value for min is internally set == max.
*   scheduler group/profile processing
* PFC enable - list of queues for which to enable PFC.

* main.cpp
*    changed log level DEBUG back to INFO
* portsqosorch.cpp, Port::getQueue
* portsqosorch.cpp, Port::getPG

* Samples
*  Added QOS and Buffer related configuration samples under swssconfig/sample.

* Schema
*	WRED schema changes
* 		Added WRED red max threshold support
*		Added ECN mode support

signed-off-by: hrachya@mellanox.com

### CODE REVIEW

File: orchagent/bufferorch.cpp [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+
+using namespace std;
+
+type_map BufferOrch::m_buffer_type_maps = {
+    {APP_BUFFER_POOL_TABLE_NAME,    new object_map()},
+    {APP_BUFFER_PROFILE_TABLE_NAME, new object_map()},
+    {APP_BUFFER_QUEUE_TABLE_NAME,   new object_map()},
+    {APP_BUFFER_PG_TABLE_NAME,     new object_map()},
+    {APP_BUFFER_PORT_INGRESS_PROFILE_LIST_NAME,     new object_map()},
+    {APP_BUFFER_PORT_EGRESS_PROFILE_LIST_NAME,     new object_map()}
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: leave only one space between

File: orchagent/bufferorch.cpp [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+            return task_process_status::task_failed;
+        }
+    }
+    return task_process_status::task_success;
+}
+
+void BufferOrch::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+    if (consumer.m_toSync.empty())
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: this can be removed

File: orchagent/bufferorch.cpp [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+task_process_status BufferOrch::processBufferProfile(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+    sai_status_t            sai_status;
+    sai_object_id_t         sai_object = SAI_NULL_OBJECT_ID;
+    auto                    it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple  tuple = it->second;
+    string                  map_type_name = consumer.m_consumer->getTableName();
+    string                  object_name = kfvKey(tuple);
+    string                  op = kfvOp(tuple);
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: Just use one space in between

File: orchagent/bufferorch.cpp [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+            sai_attribute_t attr;
+            if (fvField(*i) == buffer_pool_field_name)
+            {
+                sai_object_id_t sai_pool;
+                ref_resolve_status resolve_result = resolveFieldRefValue(m_buffer_type_maps, buffer_pool_field_name, tuple, sai_pool);
+                if (ref_resolve_status::success != resolve_result)
+                {
+                    if(ref_resolve_status::not_resolved == resolve_result)
+                    {
+                        SWSS_LOG_ERROR("Missing or invalid pool reference specified\n");
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: ERROR or WARNING

File: orchagent/bufferorch.cpp [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+    sai_status_t            sai_status;
+    sai_object_id_t         sai_object = SAI_NULL_OBJECT_ID;
+    auto                    it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple  tuple = it->second;
+    string                  map_type_name = consumer.m_consumer->getTableName();
+    string                  object_name = kfvKey(tuple);
+    string                  op = kfvOp(tuple);
+
+    SWSS_LOG_DEBUG("object name:%s", object_name.c_str());
+    if (map_type_name != APP_BUFFER_POOL_TABLE_NAME)
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: these checks are not needed and can be removed

File: orchagent/bufferorch.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    virtual void doTask(Consumer& consumer);
+    void initTableHandlers();
+    task_process_status processBufferPool(Consumer &consumer);
+    task_process_status processBufferProfile(Consumer &consumer);
+    task_process_status processQueue(Consumer &consumer);
+    task_process_status processPriorityGroup(Consumer &consumer);
+    task_process_status processIngressBufferProfileList(Consumer &consumer);
+    task_process_status processEgressBufferProfileList(Consumer &consumer);
+private:
```
Comment (stcheng) [inline orchagent/bufferorch.h]: this private is not needed.

File: orchagent/portsorch.cpp [hunk 7]
```diff
@@ -19,6 +19,80 @@ extern sai_hostif_api_t* sai_hostif_api;
 #define VLAN_PREFIX         "Vlan"
 #define DEFAULT_VLAN_ID     1
 
+namespace swss {
```
Comment (stcheng) [inline orchagent/portsorch.cpp]: no need namespace here

File: orchagent/qosorch.cpp [hunk 8]
```diff
@@ -0,0 +1,1368 @@
...
+{
+    SWSS_LOG_ENTER();
+    if (tableName != APP_DSCP_TO_TC_MAP_TABLE_NAME)
+    {
+        SWSS_LOG_ERROR("invalid table type passed in %s, expected:%s\n", tableName.c_str(), APP_DSCP_TO_TC_MAP_TABLE_NAME);
+        return false;
+    }
+    return true;
+}
+bool DscpToTcMapHandler::convertFieldValuesToAttributes(KeyOpFieldsValuesTuple &tuple, vector<sai_attribute_t> &attributes)
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: add new lines between functions

File: orchagent/qosorch.cpp [hunk 9]
```diff
@@ -0,0 +1,1368 @@
...
+    {
+        SWSS_LOG_ERROR("invalid table type passed in %s, expected:%s\n", tableName.c_str(), APP_TC_TO_QUEUE_MAP_TABLE_NAME);
+        return false;
+    }
+    return true;
+}
+bool TcToQueueMapHandler::convertFieldValuesToAttributes(KeyOpFieldsValuesTuple &tuple, vector<sai_attribute_t> &attributes)
+{
+    SWSS_LOG_ENTER();
+    sai_attribute_t     list_attr;
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: single space

File: orchagent/qosorch.cpp [hunk 10]
```diff
@@ -0,0 +1,1368 @@
...
+    list_attr.value.qosmap.count = tc_map_list.count;
+    list_attr.value.qosmap.list = tc_map_list.list;
+    attributes.push_back(list_attr);
+    return true;
+}
+
+bool TcToQueueMapHandler::modifyQosMap(sai_object_id_t sai_object, vector<sai_attribute_t> &attributes)
+{
+    SWSS_LOG_ENTER();
+    sai_status_t sai_status = sai_qos_map_api->set_qos_map_attribute(sai_object, &attributes[0]);
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: why only modify the first element of the vector

File: orchagent/qosorch.cpp [hunk 11]
```diff
@@ -0,0 +1,1368 @@
...
+    TcToQueueMapHandler tc_queue_handler;
+    return tc_queue_handler.processWorkItem(consumer);
+}
+
+void WredMapHandler::freeAttribResources(vector<sai_attribute_t> &attributes)
+{
+    SWSS_LOG_ENTER();
+}
+
+bool WredMapHandler::isValidTable(string &tableName)
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: this check is not needed

File: orchagent/qosorch.cpp [hunk 12]
```diff
@@ -0,0 +1,1368 @@
...
+        return false;
+    }
+    return true;
+}
+
+bool WredMapHandler::convertEcnMode(string str, sai_ecn_mark_mode_t &ecn_val)
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_DEBUG("input:%s", str.c_str());
+    if(str == "ecn_none")
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: construct a map for this in the qosorch.h with string mapping to SAI enums

File: orchagent/qosorch.cpp [hunk 13]
```diff
@@ -0,0 +1,1368 @@
...
+        if (sai_status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to set wred profile attribute, id:%d, status:%d", attr.id, sai_status);
+            return false;
+        }
+    }
+    return true;
+}
+
+sai_object_id_t WredMapHandler::addQosMap(vector<sai_attribute_t> &attribs_in)
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: add const to the parameter instead of using suffix _in

File: orchagent/qosorch.cpp [hunk 14]
```diff
@@ -0,0 +1,1368 @@
...
+    list_attr.value.qosmap.count = tc_to_pg_map_list.count;
+    list_attr.value.qosmap.list = tc_to_pg_map_list.list;
+    attributes.push_back(list_attr);
+    return true;
+}
+
+void TcToPgHandler::freeAttribResources(vector<sai_attribute_t> &attributes)
+{
+    SWSS_LOG_ENTER();
+    delete[] attributes[0].value.qosmap.list;
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: only free the first element?

File: orchagent/qosorch.cpp [hunk 15]
```diff
@@ -0,0 +1,1368 @@
...
+}
+
+bool TcToPgHandler::modifyQosMap(sai_object_id_t sai_object, vector<sai_attribute_t> &attributes)
+{
+    SWSS_LOG_ENTER();
+    sai_status_t sai_status = sai_qos_map_api->set_qos_map_attribute(sai_object, &attributes[0]);
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: only set the first element?

File: orchagent/qosorch.h [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+class QosMapHandler
+{
+public:
+    task_process_status processWorkItem(Consumer& consumer);
+    virtual bool isValidTable(string &tableName) = 0;
+    virtual bool convertFieldValuesToAttributes(KeyOpFieldsValuesTuple &tuple, vector<sai_attribute_t> &attributes) = 0;
+    virtual void freeAttribResources(vector<sai_attribute_t> &attributes) = 0;
+    virtual bool modifyQosMap(sai_object_id_t, vector<sai_attribute_t> &attributes) = 0;
+    virtual sai_object_id_t addQosMap(vector<sai_attribute_t> &attributes) = 0;
+    virtual bool removeQosMap(sai_object_id_t sai_object) = 0;
```
Comment (stcheng) [inline orchagent/qosorch.h]: most of the functions here need not to be pure virtual

File: orchagent/portsorch.cpp (hunk 7 above)
Comment (hrachya-mellanox) [inline orchagent/portsorch.cpp]: Port class is member of swss namespace. As we've agreed in past, this code was placed in this file.

File: orchagent/bufferorch.cpp [hunk 17]
```diff
@@ -0,0 +1,596 @@
+#include "sai.h"
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: remove this

File: orchagent/bufferorch.cpp [hunk 18]
```diff
@@ -0,0 +1,596 @@
...
+#include "tokenize.h"
+
+#include "bufferorch.h"
+#include "logger.h"
+
+#include <sstream>
+#include <iostream>
+
+extern sai_port_api_t   *sai_port_api;
+extern sai_queue_api_t  *sai_queue_api;
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: one space between

File: orchagent/bufferorch.cpp [hunk 19]
```diff
@@ -0,0 +1,596 @@
...
+extern sai_switch_api_t *sai_switch_api;
+extern sai_buffer_api_t *sai_buffer_api;
+
+using namespace std;
+
+type_map BufferOrch::m_buffer_type_maps = {
+    {APP_BUFFER_POOL_TABLE_NAME, new object_map()},
+    {APP_BUFFER_PROFILE_TABLE_NAME,  new object_map()},
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: one space between

File: orchagent/bufferorch.cpp [hunk 20]
```diff
@@ -0,0 +1,596 @@
...
+    SWSS_LOG_ENTER();
+    m_bufferHandlerMap.insert(buffer_handler_pair(APP_BUFFER_POOL_TABLE_NAME, &BufferOrch::processBufferPool));
+    m_bufferHandlerMap.insert(buffer_handler_pair(APP_BUFFER_PROFILE_TABLE_NAME, &BufferOrch::processBufferProfile));
+    m_bufferHandlerMap.insert(buffer_handler_pair(APP_BUFFER_QUEUE_TABLE_NAME, &BufferOrch::processQueue));
+    m_bufferHandlerMap.insert(buffer_handler_pair(APP_BUFFER_PG_TABLE_NAME, &BufferOrch::processPriorityGroup));
+    m_bufferHandlerMap.insert(buffer_handler_pair(APP_BUFFER_PORT_INGRESS_PROFILE_LIST_NAME, &BufferOrch::processIngressBufferProfileList));
+    m_bufferHandlerMap.insert(buffer_handler_pair(APP_BUFFER_PORT_EGRESS_PROFILE_LIST_NAME, &BufferOrch::processEgressBufferProfileList));
+}
+
+task_process_status BufferOrch::processBufferPool(Consumer &consumer)
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: why the parameter is Consumer & instead of the KeyOpFieldsValuesTuple?

File: orchagent/bufferorch.cpp [hunk 21]
```diff
@@ -0,0 +1,596 @@
...
+    sai_status_t sai_status;
+    sai_object_id_t sai_object = SAI_NULL_OBJECT_ID;
+    auto it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple tuple = it->second;
+    string map_type_name = consumer.m_consumer->getTableName();
+    string object_name = kfvKey(tuple);
+    string op = kfvOp(tuple);
+
+    SWSS_LOG_DEBUG("object name:%s", object_name.c_str());
+    if (map_type_name != APP_BUFFER_PROFILE_TABLE_NAME)
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: remove this

File: orchagent/bufferorch.cpp [hunk 22]
```diff
@@ -0,0 +1,596 @@
...
+    sai_uint32_t range_low, range_high;
+
+    SWSS_LOG_DEBUG("Processing:%s", key.c_str());
+    tokens = tokenize(key, delimiter);
+    if (tokens.size() != 2)
+    {
+        SWSS_LOG_ERROR("malformed key:%s. Must contain 2 tokens\n", key.c_str());
+        return task_process_status::task_invalid_entry;
+    }
+    if (consumer.m_consumer->getTableName() != APP_BUFFER_QUEUE_TABLE_NAME)
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: remove this

File: orchagent/bufferorch.cpp [hunk 23]
```diff
@@ -0,0 +1,596 @@
...
+    sai_uint32_t range_low, range_high;
+
+    SWSS_LOG_DEBUG("processing:%s", key.c_str());
+    tokens = tokenize(key, delimiter);
+    if (tokens.size() != 2)
+    {
+        SWSS_LOG_ERROR("malformed key:%s. Must contain 2 tokens\n", key.c_str());
+        return task_process_status::task_invalid_entry;
+    }
+    if (consumer.m_consumer->getTableName() != APP_BUFFER_PG_TABLE_NAME)
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: remove this

File: orchagent/bufferorch.cpp [hunk 24]
```diff
@@ -0,0 +1,596 @@
...
+task_process_status BufferOrch::processEgressBufferProfileList(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+    auto it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple tuple = it->second;
+    Port port;
+    string key = kfvKey(tuple);
+    string op = kfvOp(tuple);
+    SWSS_LOG_DEBUG("processing:%s", key.c_str());
+    if (consumer.m_consumer->getTableName() != APP_BUFFER_PORT_EGRESS_PROFILE_LIST_NAME)
```
Comment (stcheng) [inline orchagent/bufferorch.cpp]: remove this

File: orchagent/qosorch.cpp [hunk 25]
```diff
@@ -0,0 +1,1147 @@
...
+
+    ref_resolve_status  resolve_result;
+    // sample "QUEUE_TABLE:ETHERNET4:1"
+    tokens = tokenize(key, delimiter);
+    if (tokens.size() != 2)
+    {
+        SWSS_LOG_ERROR("malformed key:%s. Must contain 2 tokens\n", key.c_str());
+        return task_process_status::task_invalid_entry;
+    }
+    if (consumer.m_consumer->getTableName() != APP_QUEUE_TABLE_NAME)
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: remove this

File: orchagent/qosorch.cpp [hunk 26]
```diff
@@ -0,0 +1,1147 @@
...
+            {
+                sai_uint8_t q_val = stoi(q_ind);
+                pfc_enable |= (1 << q_val);
+            }
+            break;
+        }
+    }
+
+    //"PORT_QOS_MAP_TABLE:ETHERNET4
+    if (consumer.m_consumer->getTableName() != APP_PORT_QOS_MAP_TABLE_NAME)
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: remove this

File: orchagent/orch.cpp [hunk 27]
```diff
@@ -127,3 +206,97 @@ void Orch::dumpTuple(Consumer &consumer, KeyOpFieldsValuesTuple &tuple)
...
+    }
+    if(0 == count)
+    {
+        SWSS_LOG_NOTICE("field with name:%s not found\n", field_name.c_str());
+        return ref_resolve_status::field_not_found;
+    }
+    return ref_resolve_status::success;
+}
+
+vector<string> Orch::parseNameArray(const string &input)
```
Comment (stcheng) [inline orchagent/orch.cpp]: Please remove this function, the tokenize() function supports single token case (find == string::npos)

File: orchagent/orch.cpp [hunk 28]
```diff
@@ -127,3 +206,97 @@ void Orch::dumpTuple(Consumer &consumer, KeyOpFieldsValuesTuple &tuple)
     }
     SWSS_LOG_DEBUG("%s\n", debug_msg.c_str());
 }
+
+ref_resolve_status Orch::resolveFieldRefArray(
+    type_map                    &type_maps,
```
Comment (stcheng) [inline orchagent/orch.cpp]: single space in betwen

File: orchagent/orch.cpp [hunk 29]
```diff
@@ -103,17 +101,98 @@ bool Orch::execute(string tableName)
...
+        return false;
+    }
+    type_name   = tokens[0];
+    object_name = tokens[1];
+    SWSS_LOG_DEBUG("parsed: type_name:%s, object_name:%s", type_name.c_str(), object_name.c_str());
+    return true;
+}
+
+ref_resolve_status Orch::resolveFieldRefValue(
+    type_map                &type_maps,
```
Comment (stcheng) [inline orchagent/orch.cpp]: single space in between

File: orchagent/orch.cpp [hunk 30]
```diff
@@ -103,17 +101,98 @@ bool Orch::execute(string tableName)
...
+        return false;
+    }
+    auto obj_map = type_maps[tokens[0]];
+    auto obj_it = obj_map->find(tokens[1]);
+    if(obj_it == obj_map->end())
+    {
+        SWSS_LOG_ERROR("map:%s does not contain object with name:%s\n", tokens[0].c_str(), tokens[1].c_str());
+        return false;
+    }
+    type_name   = tokens[0];
```
Comment (stcheng) [inline orchagent/orch.cpp]: single space in between

File: orchagent/orch.h [hunk 31]
```diff
@@ -27,41 +31,65 @@ typedef enum
...
     DBConnector *m_db;
-
 protected:
     ConsumerMap m_consumerMap;
 
+    /* Run doTask against a specific consumer */
+    virtual void doTask(Consumer &consumer) = 0;
+    void dumpTuple(Consumer &consumer, KeyOpFieldsValuesTuple &tuple);
+    ref_resolve_status resolveFieldRefValue(
+        type_map                &type_maps,
```
Comment (stcheng) [inline orchagent/orch.h]: put in one line without multiple spaces. the variable names can be omitted here

File: orchagent/orchdaemon.cpp [hunk 32]
```diff
@@ -43,9 +42,32 @@ bool OrchDaemon::init()
     RouteOrch *route_orch = new RouteOrch(m_applDb, APP_ROUTE_TABLE_NAME, ports_orch, neigh_orch);
     CoppOrch  *copp_orch  = new CoppOrch(m_applDb, APP_COPP_TABLE_NAME);
     TunnelDecapOrch *tunnel_decap_orch = new TunnelDecapOrch(m_applDb, APP_TUNNEL_DECAP_TABLE_NAME);
-    
-    m_orchList = { ports_orch, intfs_orch, neigh_orch, route_orch, copp_orch, tunnel_decap_orch };
+
+    std::vector<std::string> qos_tables = {
```
Comment (stcheng) [inline orchagent/orchdaemon.cpp]: remove std::

File: orchagent/orchdaemon.cpp [hunk 33]
```diff
@@ -43,9 +42,32 @@ bool OrchDaemon::init()
...
+        APP_QUEUE_TABLE_NAME,
+        APP_PORT_QOS_MAP_TABLE_NAME,
+        APP_WRED_PROFILE_TABLE_NAME,
+        APP_TC_TO_PRIORITY_GROUP_MAP_NAME,
+        APP_PFC_PRIORITY_TO_PRIORITY_GROUP_MAP_NAME,
+        APP_PFC_PRIORITY_TO_QUEUE_MAP_NAME
+    };
+    QosOrch *qos_orch = new QosOrch(m_applDb, qos_tables, ports_orch);
+
+    std::vector<std::string> buffer_tables = {
```
Comment (stcheng) [inline orchagent/orchdaemon.cpp]: remove std::

File: orchagent/portsorch.cpp [hunk 34]
```diff
@@ -537,12 +537,77 @@ void PortsOrch::doTask(Consumer &consumer)
...
 {
     SWSS_LOG_ENTER();
 
     SWSS_LOG_NOTICE("Initializing port alias:%s pid:%llx\n", p.m_alias.c_str(), p.m_port_id);
 
+    if (!initializePriorityGroups(p))
+    {
+        SWSS_LOG_ERROR("Failed to initialize port priority groups pid:%llx alias:%s\n", p.m_port_id, p.m_alias.c_str());
+        return false;
+    }
```
Comment (stcheng) [inline orchagent/portsorch.cpp]: new line in between

File: orchagent/qosorch.cpp [hunk 35]
```diff
@@ -0,0 +1,1147 @@
...
+#include "tokenize.h"
+#include "qosorch.h"
+#include "logger.h"
+
+#include <sstream>
+#include <iostream>
+
+using namespace std;
+
+extern sai_port_api_t               *sai_port_api;
```
Comment (stcheng) [inline orchagent/qosorch.cpp]: one space in between

File: orchagent/portsorch.cpp [hunk 36]
```diff
@@ -537,12 +537,77 @@ void PortsOrch::doTask(Consumer &consumer)
...
+        SWSS_LOG_ERROR("fail to call sai_port_api->get_port_attribute: port:%s, status:%d", port.m_alias.c_str(), status);
+        return false;
+    }
+    return true;
+}
+
+bool PortsOrch::initializePriorityGroups(Port &port)
+{
+    SWSS_LOG_ENTER();
+    sai_attribute_t  attr;
```
Comment (stcheng) [inline orchagent/portsorch.cpp]: one space in between

File: orchagent/bufferorch.cpp [hunk 37]
```diff
@@ -0,0 +1,575 @@
...
+
+extern sai_port_api_t *sai_port_api;
+extern sai_queue_api_t *sai_queue_api;
+extern sai_switch_api_t *sai_switch_api;
+extern sai_buffer_api_t *sai_buffer_api;
+
+using namespace std;
+
+type_map BufferOrch::m_buffer_type_maps = {
+    {APP_BUFFER_POOL_TABLE_NAME, new object_map()},
```
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: How to handle exception thrown inside new operator?

File: orchagent/bufferorch.cpp [hunk 38]
```diff
@@ -0,0 +1,575 @@
...
+    }
+    SWSS_LOG_DEBUG("processing command:%s", op.c_str());
+    if (op == SET_COMMAND)
+    {
+        vector<sai_attribute_t> attribs;
+        for (auto i = kfvFieldsValues(tuple).begin(); i != kfvFieldsValues(tuple).end(); i++)
+        {
+            SWSS_LOG_DEBUG("field:%s, value:%s", fvField(*i).c_str(), fvValue(*i).c_str());
+            sai_attribute_t attr;
+            if (fvField(*i) == buffer_size_field_name)
```
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: switch case block?

Comment (hrachya-mellanox) [inline orchagent/bufferorch.cpp]: Does C++ allow strings as switch operator arguments?

File: orchagent/bufferorch.cpp (hunk 37 above)
Comment (hrachya-mellanox) [inline orchagent/bufferorch.cpp]: exception thrown in global scope will terminate executable on startup.

File: orchagent/bufferorch.cpp (hunk 38 above)
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: Got it. Feel free to ignore previous comment.

**Comment by hrachya-mellanox** [comment]
Hi,
Can we please merge the change since all comments have been addressed?

**Pull request merged**